armory/utils/export.py

```diff
@@ -1,74 +1,51 @@
 import os
 import logging
+import abc
 import numpy as np
 import ffmpeg
 import pickle
 import time
 from PIL import Image, ImageDraw
 from scipy.io import wavfile
 
-from armory.data.datasets import ImageContext, VideoContext, AudioContext, So2SatContext
-
 
 logger = logging.getLogger(__name__)
 
 
 class SampleExporter:
-    def __init__(self, base_output_dir, context, num_samples):
+    def __init__(self, base_output_dir, num_samples):
         self.base_output_dir = base_output_dir
-        self.context = context
         self.num_samples = num_samples
         self.saved_samples = 0
         self.output_dir = None
         self.y_dict = {}
 
-        if isinstance(self.context, VideoContext):
-            self.export_fn = self._export_video
-        elif isinstance(self.context, ImageContext):
-            self.export_fn = self._export_images
-        elif isinstance(self.context, AudioContext):
-            self.export_fn = self._export_audio
-        elif isinstance(self.context, So2SatContext):
-            self.export_fn = self._export_so2sat
-        else:
-            raise TypeError(
-                f"Expected VideoContext, ImageContext, AudioContext, or So2SatContext, got {type(self.context)}"
-            )
         self._make_output_dir()
 
     def export(
```

Comment:
Could we modify the export function so that it could also just return the annotated image/video/audio? This would be very nice for jupyter notebooks / interactive use. This could either be its own function call, like "annotate", that shares the same underlying code, or a flag that gets passed in like `return_sample=True` and skips the last step of the saving process.

Reply:
this seems straightforward for image scenarios: return the PIL object. I'm less clear on what this would look like for video and audio though. It doesn't seem particularly useful to return the np array since this can already be accessed via `self.x` or `self.x_adv`.

Reply:
For so2sat, I would return a list of PIL objects.
For video, it looks like you're separately saving frames. I would return either that set of PIL images, or the annotated numpy "pixels" arrays.
For audio,  it needs to be mapped back to the discrete np.int16 input plus sample rate. Maybe just return that? There's not a ton to do in the audio space, at least at the moment.

What I don't know yet (aside from PIL) is what is the easiest way to show a video or play audio in jupyter.

Reply:
I added a handful of commits in an attempt to address this point. They add a `get_sample()` method to each Exporter which can be used to return the PIL image, list of PIL frames, etc. depending on the scenario. This `get_sample()` method is then used internally by each Exporter's `_export()` method (the lone exception being Audio)

The `AudioExporter`'s `get_sample()` is unique in that it requires a `dataset_context` arg since `context.quantization` is needed to scale back the sample. However, I think I noticed in `datasets.py` [here](https://github.com/twosixlabs/armory/blob/develop/armory/data/datasets.py#L735) that  2 **15 is always used similar to how 255 is always used for images, so perhaps this is unnecessary?